# First Look to the Data and Cleaning & Exploring

Water is one of the main sources for people and also countries. Even clean water is worthless. Tanzania is a developing country which has clean water problem. According to [worldometers.info](https://www.worldometers.info/world-population/tanzania-population/), Tanzanian population is 59,353,795 and it has to provide more than 59 million people to clean water. There are many water wells which already established, but also there are many useless wells or some are in need of repair. 

Drivendata has begun a competition 'Pump It Up' to point this problem. With the predictive model, people can understand which water points are functional, nonfunctional and functional but it needs to repair. This model can help the Tanzanian government to find likely maintenance needy wells or give an useful information for future wells. 

Our aim in the project to build a model which tells the status of the water points (as functional, nonfunctional, functional but needs repair). With this model, we will help to the Tanzanian authorities how to use water sources as a productive way. It also helps the investment of the government on wells. 

## Aim of This Notebook 

Aim of this notebook is data understanding, preparation and exploration. In this notebook, we will download data, try to understand the columns meanings. We will also discover what kind of data cleaning is required. With the help of visualizations, we will look at the some features. Then, we will prepare our data to modeling. We will drop some duplicated or uncessary informations or handle null or meaningless valeus.

At the end of the notebook, all explorations, which are obtained by the first look of the data, can be found as a list.

## Data Source

The original data can be obtained by the [DrivenData 'Pump it Up: Data Mining the Water Table' competition](https://www.drivendata.org/competitions/7/pump-it-up-data-mining-the-water-table/page/23/). Basically, there are 4 different data sets; submission format, training set, test set and train labels set which contains status of wells.  With given training set and labels set, competitors are wanted to build predictive model and apply it to test set to determine status of the wells and submit. 

In this project, we will use train set and train label set. Train set has 59400 water points data with 40 features. Train labels data has 59400 same water points with train set but just has information about id of these points and status of them.


## Importing Necessary Libraries 

In [1]:
import numpy as np      # To use np.arrays
import pandas as pd     # To use dataframes

# To plot
import matplotlib.pyplot as plt  
%matplotlib inline        
import seaborn as sns

In [2]:
pd.options.display.max_columns=100 # To see the hidden columns in dataframe

In [3]:
df_train_set = pd.read_csv('training_set_values.csv') # train set data

In [4]:
df_train_labels = pd.read_csv('training_set_labels.csv') #train set labels data

# First Look to the Data and Merging Two Dataframes

In [5]:
df_train_labels

id    status_group
0      69572      functional
1       8776      functional
2      34310      functional
3      67743  non functional
4      19728      functional
...      ...             ...
59395  60739      functional
59396  27263      functional
59397  37057      functional
59398  31282      functional
59399  26348      functional

[59400 rows x 2 columns]

In [6]:
df_train_set

id  amount_tsh date_recorded           funder  gps_height  \
0      69572      6000.0    2011-03-14            Roman        1390   
1       8776         0.0    2013-03-06          Grumeti        1399   
2      34310        25.0    2013-02-25     Lottery Club         686   
3      67743         0.0    2013-01-28           Unicef         263   
4      19728         0.0    2011-07-13      Action In A           0   
...      ...         ...           ...              ...         ...   
59395  60739        10.0    2013-05-03  Germany Republi        1210   
59396  27263      4700.0    2011-05-07      Cefa-njombe        1212   
59397  37057         0.0    2011-04-11              NaN           0   
59398  31282         0.0    2011-03-08            Malec           0   
59399  26348         0.0    2011-03-23       World Bank         191   

          installer  longitude   latitude              wpt_name  num_private  \
0             Roman  34.938093  -9.856322                  none            0   
1           GRUMETI  34.698766  -2.147466              Zahanati            0   
2      World vision  37.460664  -3.821329           Kwa Mahundi            0   
3            UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
4           Artisan  31.130847  -1.825359               Shuleni            0   
...             ...        ...        ...                   ...          ...   
59395           CES  37.169807  -3.253847   Area Three Namba 27            0   
59396          Cefa  35.249991  -9.070629     Kwa Yahona Kuvala            0   
59397           NaN  34.017087  -8.750434               Mashine            0   
59398          Musa  35.861315  -6.378573                Mshoro            0   
59399         World  38.104048  -6.747464       Kwa Mzee Lugawa            0   

                         basin    subvillage       region  region_code  \
0                   Lake Nyasa      Mnyusi B       Iringa           11   
1                Lake Victoria       Nyamara         Mara           20   
2                      Pangani       Majengo      Manyara           21   
3      Ruvuma / Southern Coast    Mahakamani       Mtwara           90   
4                Lake Victoria    Kyanyamisa       Kagera           18   
...                        ...           ...          ...          ...   
59395                  Pangani      Kiduruni  Kilimanjaro            3   
59396                   Rufiji      Igumbilo       Iringa           11   
59397                   Rufiji     Madungulu        Mbeya           12   
59398                   Rufiji        Mwinyi       Dodoma            1   
59399              Wami / Ruvu  Kikatanyemba     Morogoro            5   

       district_code             lga              ward  population  \
0                  5          Ludewa          Mundindi         109   
1                  2       Serengeti             Natta         280   
2                  4       Simanjiro           Ngorika         250   
3                 63        Nanyumbu          Nanyumbu          58   
4                  1         Karagwe        Nyakasimbi           0   
...              ...             ...               ...         ...   
59395              5             Hai  Masama Magharibi         125   
59396              4          Njombe            Ikondo          56   
59397              7         Mbarali           Chimala           0   
59398              4        Chamwino      Mvumi Makulu           0   
59399              2  Morogoro Rural        Ngerengere         150   

      public_meeting              recorded_by scheme_management  \
0               True  GeoData Consultants Ltd               VWC   
1                NaN  GeoData Consultants Ltd             Other   
2               True  GeoData Consultants Ltd               VWC   
3               True  GeoData Consultants Ltd               VWC   
4               True  GeoData Consultants Ltd               NaN   
...              ...                      ...               ... 

The index numbers in two data sets look same but it is impossible to check the nearly 60000 points. So, to make sure we assign data sets 'id' as index and after that merge them. 

In [7]:
df_train_set.set_index('id',inplace=True) # setting id as an index to train set

In [8]:
df_train_labels.set_index('id',inplace=True) # setting id as an index to train labels

In [9]:
df = pd.merge(df_train_labels, df_train_set, how = 'inner', left_index = True,right_index=True) #merging two data sets

In [10]:
df.reset_index(inplace=True) # setting a new index 

# Looking at Columns

In [11]:
df.info() #to see the types of the columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 41 columns):
id                       59400 non-null int64
status_group             59400 non-null object
amount_tsh               59400 non-null float64
date_recorded            59400 non-null object
funder                   55765 non-null object
gps_height               59400 non-null int64
installer                55745 non-null object
longitude                59400 non-null float64
latitude                 59400 non-null float64
wpt_name                 59400 non-null object
num_private              59400 non-null int64
basin                    59400 non-null object
subvillage               59029 non-null object
region                   59400 non-null object
region_code              59400 non-null int64
district_code            59400 non-null int64
lga                      59400 non-null object
ward                     59400 non-null object
population               59400 non-null int64
p

In [12]:
df.isna().sum() # to see the null values

id                           0
status_group                 0
amount_tsh                   0
date_recorded                0
funder                    3635
gps_height                   0
installer                 3655
longitude                    0
latitude                     0
wpt_name                     0
num_private                  0
basin                        0
subvillage                 371
region                       0
region_code                  0
district_code                0
lga                          0
ward                         0
population                   0
public_meeting            3334
recorded_by                  0
scheme_management         3877
scheme_name              28166
permit                    3056
construction_year            0
extraction_type              0
extraction_type_group        0
extraction_type_class        0
management                   0
management_group             0
payment                      0
payment_type                 0
water_qu

In [13]:
df.describe() # to see numeric columns detailed

id     amount_tsh    gps_height     longitude      latitude  \
count  59400.000000   59400.000000  59400.000000  59400.000000  5.940000e+04   
mean   37115.131768     317.650385    668.297239     34.077427 -5.706033e+00   
std    21453.128371    2997.574558    693.116350      6.567432  2.946019e+00   
min        0.000000       0.000000    -90.000000      0.000000 -1.164944e+01   
25%    18519.750000       0.000000      0.000000     33.090347 -8.540621e+00   
50%    37061.500000       0.000000    369.000000     34.908743 -5.021597e+00   
75%    55656.500000      20.000000   1319.250000     37.178387 -3.326156e+00   
max    74247.000000  350000.000000   2770.000000     40.345193 -2.000000e-08   

        num_private   region_code  district_code    population  \
count  59400.000000  59400.000000   59400.000000  59400.000000   
mean       0.474141     15.297003       5.629747    179.909983   
std       12.236230     17.587406       9.633649    471.482176   
min        0.000000      1.000000       0.000000      0.000000   
25%        0.000000      5.000000       2.000000      0.000000   
50%        0.000000     12.000000       3.000000     25.000000   
75%        0.000000     17.000000       5.000000    215.000000   
max     1776.000000     99.000000      80.000000  30500.000000   

       construction_year  
count       59400.000000  
mean         1300.652475  
std           951.620547  
min             0.000000  
25%             0.000000  
50%          1986.000000  
75%          2004.000000  
max          2013.000000

In [14]:
df['status_group'].value_counts()

functional                 32259
non functional             22824
functional needs repair     4317
Name: status_group, dtype: int64

We have highly imbalanced target values. Also, there are some null values and some columns which contain same information in the data set. Now, we will drop them because the same values or dublicated values do not affect our target, and when we simplify our data we can run our models easier. 

### scheme_management / management / management_group columns

In [15]:
df['scheme_management'].value_counts()

VWC                 36793
WUG                  5206
Water authority      3153
WUA                  2883
Water Board          2748
Parastatal           1680
Private operator     1063
Company              1061
Other                 766
SWC                    97
Trust                  72
None                    1
Name: scheme_management, dtype: int64

In [16]:
df['management'].value_counts()

vwc                 40507
wug                  6515
water board          2933
wua                  2535
private operator     1971
parastatal           1768
water authority       904
other                 844
company               685
unknown               561
other - school         99
trust                  78
Name: management, dtype: int64

These two columns has nearly same information. Because scheme_management represents who operates the water point, 'management' represents how the water point is managed. There are 3877 null values in 'scheme_management' column so we prefer 'management' column to keep. The column 'management_group' is also keep similar information that how the water point is managed.

In [17]:
df['management_group'].value_counts()

user-group    52490
commercial     3638
parastatal     1768
other           943
unknown         561
Name: management_group, dtype: int64

In [18]:
df.loc[df['management_group']=='user-group']['management'].value_counts() # to find sub-groups of this column

vwc            40507
wug             6515
water board     2933
wua             2535
Name: management, dtype: int64

To find the subgroups of 'management_group' column, we checked the 'user-group' values and saw that this column is just the grouped version of 'management'. Although 'management' has more detailed values, we decided to drop 'management_group' column. Also, to remember and check the sub-groups of the 'management_group' column, we grouped it below and saw the numbers of sub-groups (management column).

In [19]:
df.groupby(['management_group','management']).count() 
# to see how many sub-groups have in management group according to management column

id  status_group  amount_tsh  \
management_group management                                          
commercial       company             685           685         685   
                 private operator   1971          1971        1971   
                 trust                78            78          78   
                 water authority     904           904         904   
other            other               844           844         844   
                 other - school       99            99          99   
parastatal       parastatal         1768          1768        1768   
unknown          unknown             561           561         561   
user-group       vwc               40507         40507       40507   
                 water board        2933          2933        2933   
                 wua                2535          2535        2535   
                 wug                6515          6515        6515   

                                   date_recorded  funder  gps_height  \
management_group management                                            
commercial       company                     685     663         685   
                 private operator           1971    1957        1971   
                 trust                        78      78          78   
                 water authority             904     836         904   
other            other                       844     837         844   
                 other - school               99      99          99   
parastatal       parastatal                 1768    1624        1768   
unknown          unknown                     561     533         561   
user-group       vwc                       40507   37632       40507   
                 water board                2933    2715        2933   
                 wua                        2535    2308        2535   
                 wug                        6515    6483        6515   

                                   installer  longitude  latitude  wpt_name  \
management_group management                                                   
commercial       company                 663        685       685       685   
                 private operator       1959       1971      1971      1971   
                 trust                    78         78        78        78   
                 water authority         836        904       904       904   
other            other                   831        844       844       844   
                 other - school           99         99        99        99   
parastatal       parastatal             1626       1768      1768      1768   
unknown          unknown                 527        561       561       561   
user-group       vwc                   37630      40507     40507     40507   
                 water board            2714       2933      2933      2933   
                 wua                    2309       2535      2535      2535   
                 wug                    6473       6515      6515      6515   

                                   num_private  basin  subvillage  region  \
management_group management                                                 
commercial       company                   685    685         685     685   
                 private operator         1971   1971        1932    1971   
                 trust                      78     78          78      78   
                 water authority           904    904         895     904   
other            other                     844    844         839     844   
                 other - school             99     99          99      99   
parastatal       parastatal               1768   1768        1768    1768   
unknown          unknown                   561    561         561     561   
user-group       vwc                     40507  40507       40189   40507   
                 water board              2933   2933        2933    2933   
                 

### quantity / quantity_group columns

In [20]:
df['quantity'].value_counts()

enough          33186
insufficient    15129
dry              6246
seasonal         4050
unknown           789
Name: quantity, dtype: int64

In [21]:
df['quantity_group'].value_counts()

enough          33186
insufficient    15129
dry              6246
seasonal         4050
unknown           789
Name: quantity_group, dtype: int64

These two columns contain same information so we decided to drop 'quantity_group' column.

### source / source_type / source_class columns

In [22]:
df['source'].value_counts()

spring                  17021
shallow well            16824
machine dbh             11075
river                    9612
rainwater harvesting     2295
hand dtw                  874
lake                      765
dam                       656
other                     212
unknown                    66
Name: source, dtype: int64

In [23]:
df['source_type'].value_counts()

spring                  17021
shallow well            16824
borehole                11949
river/lake              10377
rainwater harvesting     2295
dam                       656
other                     278
Name: source_type, dtype: int64

In [24]:
df['source_class'].value_counts()

groundwater    45794
surface        13328
unknown          278
Name: source_class, dtype: int64

In [25]:
df.groupby(['source_class','source']).count()
# to see how many sub-groups have in source_class according to source column

id  status_group  amount_tsh  \
source_class source                                                  
groundwater  hand dtw                874           874         874   
             machine dbh           11075         11075       11075   
             shallow well          16824         16824       16824   
             spring                17021         17021       17021   
surface      dam                     656           656         656   
             lake                    765           765         765   
             rainwater harvesting   2295          2295        2295   
             river                  9612          9612        9612   
unknown      other                   212           212         212   
             unknown                  66            66          66   

                                   date_recorded  funder  gps_height  \
source_class source                                                    
groundwater  hand dtw                        874     868         874   
             machine dbh                   11075   10252       11075   
             shallow well                  16824   16302       16824   
             spring                        17021   15870       17021   
surface      dam                             656     647         656   
             lake                            765     763         765   
             rainwater harvesting           2295    2099        2295   
             river                          9612    8715        9612   
unknown      other                           212     204         212   
             unknown                          66      45          66   

                                   installer  longitude  latitude  wpt_name  \
source_class source                                                           
groundwater  hand dtw                    868        874       874       874   
             machine dbh               10246      11075     11075     11075   
             shallow well              16286      16824     16824     16824   
             spring                    15870      17021     17021     17021   
surface      dam                         646        656       656       656   
             lake                        762        765       765       765   
             rainwater harvesting       2096       2295      2295      2295   
             river                      8721       9612      9612      9612   
unknown      other                       204        212       212       212   
             unknown                      46         66        66        66   

                                   num_private  basin  subvillage  region  \
source_class source                                                         
groundwater  hand dtw                      874    874         874     874   
             machine dbh                 11075  11075       10849   11075   
             shallow well                16824  16824       16817   16824   
             spring                      17021  17021       16886   17021   
surface      dam                           656    656         656     656   
             lake                          765    765         764     765   
             rainwater harvesting         2295   2295        2293    2295   
             river                        9612   9612        9612    9612   
unknown      other                         212    212         212     212   
             unknown                        66     66          66      66   

                                   region_code  district_code    lga   ward  \
source_class source                                                           
groundwater  hand dtw                      874            874    874    874   
             machine dbh                 11075          11075  11075  11075   
             shallow well                16824          16824  16824  16824   
             spring                      17021          17021  

It is obvious that these three columns keep same information. so, we decided to keep just 'source' column, because it has more detailed information and we will drop others.

### water_quality / quality_group columns

In [26]:
df['water_quality'].value_counts()

soft                  50818
salty                  4856
unknown                1876
milky                   804
coloured                490
salty abandoned         339
fluoride                200
fluoride abandoned       17
Name: water_quality, dtype: int64

In [27]:
df['quality_group'].value_counts()

good        50818
salty        5195
unknown      1876
milky         804
colored       490
fluoride      217
Name: quality_group, dtype: int64

'water_quality' column has more unique values, so we will keep 'water_quality' and drop 'quality_group'.

### payment / payment_type columns

In [28]:
df['payment'].value_counts()

never pay                25348
pay per bucket            8985
pay monthly               8300
unknown                   8157
pay when scheme fails     3914
pay annually              3642
other                     1054
Name: payment, dtype: int64

In [29]:
df['payment_type'].value_counts()

never pay     25348
per bucket     8985
monthly        8300
unknown        8157
on failure     3914
annually       3642
other          1054
Name: payment_type, dtype: int64

These two columns are same so we decided to drop one of them.

### extraction_type / extraction_type_group / extraction_type_class columns

In [30]:
df['extraction_type'].value_counts()

gravity                      26780
nira/tanira                   8154
other                         6430
submersible                   4764
swn 80                        3670
mono                          2865
india mark ii                 2400
afridev                       1770
ksb                           1415
other - rope pump              451
other - swn 81                 229
windmill                       117
india mark iii                  98
cemo                            90
other - play pump               85
walimi                          48
climax                          32
other - mkulima/shinyanga        2
Name: extraction_type, dtype: int64

In [31]:
df['extraction_type_group'].value_counts()

gravity            26780
nira/tanira         8154
other               6430
submersible         6179
swn 80              3670
mono                2865
india mark ii       2400
afridev             1770
rope pump            451
other handpump       364
other motorpump      122
wind-powered         117
india mark iii        98
Name: extraction_type_group, dtype: int64

In [32]:
df['extraction_type_class'].value_counts()

gravity         26780
handpump        16456
other            6430
submersible      6179
motorpump        2987
rope pump         451
wind-powered      117
Name: extraction_type_class, dtype: int64

In [33]:
df.groupby(['extraction_type_class','extraction_type_group']).count()
# to see how many sub-groups have in extraction_type_clas according to extraction_type_group

id  status_group  amount_tsh  \
extraction_type_class extraction_type_group                                    
gravity               gravity                26780         26780       26780   
handpump              afridev                 1770          1770        1770   
                      india mark ii           2400          2400        2400   
                      india mark iii            98            98          98   
                      nira/tanira             8154          8154        8154   
                      other handpump           364           364         364   
                      swn 80                  3670          3670        3670   
motorpump             mono                    2865          2865        2865   
                      other motorpump          122           122         122   
other                 other                   6430          6430        6430   
rope pump             rope pump                451           451         451   
submersible           submersible             6179          6179        6179   
wind-powered          wind-powered             117           117         117   

                                             date_recorded  funder  \
extraction_type_class extraction_type_group                          
gravity               gravity                        26780   24704   
handpump              afridev                         1770    1668   
                      india mark ii                   2400    2358   
                      india mark iii                    98      98   
                      nira/tanira                     8154    7899   
                      other handpump                   364     353   
                      swn 80                          3670    3596   
motorpump             mono                            2865    2577   
                      other motorpump                  122     122   
other                 other                           6430    6010   
rope pump             rope pump                        451     448   
submersible           submersible                     6179    5820   
wind-powered          wind-powered                     117     112   

                                             gps_height  installer  longitude  \
extraction_type_class extraction_type_group                                     
gravity               gravity                     26780      24714      26780   
handpump              afridev                      1770       1665       1770   
                      india mark ii                2400       2358       2400   
                      india mark iii                 98         98         98   
                      nira/tanira                  8154       7885       8154   
                      other handpump                364        354        364   
                      swn 80                       3670       3593       3670   
motorpump             mono                         2865       2578       2865   
                      other motorpump               122        122        122   
other                 other                        6430       6002       6430   
rope pump             rope pump                     451        448        451   
submersible           submersible                  6179       5816       6179   
wind-powered          wind-powered                  117        112        117   

                                             latitude  wpt_name  num_private  \
extraction_type_class extraction_type_group                                    
gravity               gravity                   26780     26780        26780   
handpump              afridev                    1770      1770         1770   
                      india mark ii              2400      2400         2400   
                      india mark iii               98        98           98   
                      nira/tanira                8154      8154         8154   
                    

It is obviously seen that these three columns keep same information. So, we decided to keep 'extraction_type_group' and drop others. Although, extraction_type has more unique values than extraction_type_group , some of these values are very small amount according to this big dataset. We prefered to use more compact one. Also, extraction_type_class contains less detail. So, extraction_type_group is chosen to keep.

### waterpoint_type / waterpoint_type_group

In [34]:
df['waterpoint_type'].value_counts()

communal standpipe             28522
hand pump                      17488
other                           6380
communal standpipe multiple     6103
improved spring                  784
cattle trough                    116
dam                                7
Name: waterpoint_type, dtype: int64

In [35]:
df['waterpoint_type_group'].value_counts()

communal standpipe    34625
hand pump             17488
other                  6380
improved spring         784
cattle trough           116
dam                       7
Name: waterpoint_type_group, dtype: int64

We decided to keep 'waterpoint_type' which contains more detail.

# Dropping Similar Columns

In [36]:
df.drop(columns=['management_group','scheme_management','quantity_group','source_class','source_type','quality_group',
                'payment_type','extraction_type_class','extraction_type', 'waterpoint_type_group'],inplace=True )

# Continue to Exploring Columns

We continue to check data for finding unnecessary or wrong values.

### construction_year column

In [37]:
df['construction_year'].value_counts()

0       20709
2010     2645
2008     2613
2009     2533
2000     2091
2007     1587
2006     1471
2003     1286
2011     1256
2004     1123
2012     1084
2002     1075
1978     1037
1995     1014
2005     1011
1999      979
1998      966
1990      954
1985      945
1980      811
1996      811
1984      779
1982      744
1994      738
1972      708
1974      676
1997      644
1992      640
1993      608
2001      540
1988      521
1983      488
1975      437
1986      434
1976      414
1970      411
1991      324
1989      316
1987      302
1981      238
1977      202
1979      192
1973      184
2013      176
1971      145
1960      102
1967       88
1963       85
1968       77
1969       59
1964       40
1962       30
1961       21
1965       19
1966       17
Name: construction_year, dtype: int64

In [38]:
df.loc[df['construction_year']!=0].describe()

id     amount_tsh    gps_height     longitude      latitude  \
count  38691.000000   38691.000000  38691.000000  38691.000000  38691.000000   
mean   37083.008736     466.457534   1002.367760     35.983262     -6.235372   
std    21420.922010    3541.036030    618.078669      2.558709      2.761317   
min        1.000000       0.000000    -63.000000     29.607122    -11.649440   
25%    18489.500000       0.000000    372.000000     34.676719     -8.755274   
50%    37078.000000       0.000000   1154.000000     36.648187     -6.064216   
75%    55514.500000     200.000000   1488.000000     37.803940     -3.650661   
max    74247.000000  350000.000000   2770.000000     40.345193     -1.042375   

        num_private   region_code  district_code    population  \
count  38691.000000  38691.000000   38691.000000  38691.000000   
mean       0.707710     15.706495       5.969786    269.799617   
std       15.083957     21.003006      10.700673    552.343746   
min        0.000000      2.000000       1.000000      0.000000   
25%        0.000000      4.000000       2.000000     30.000000   
50%        0.000000     11.000000       3.000000    150.000000   
75%        0.000000     16.000000       5.000000    305.000000   
max     1776.000000     99.000000      63.000000  30500.000000   

       construction_year  
count       38691.000000  
mean         1996.814686  
std            12.472045  
min          1960.000000  
25%          1987.000000  
50%          2000.000000  
75%          2008.000000  
max          2013.000000

In [48]:
df['construction_year'].replace(to_replace = 0, value = 2000, inplace=True)

### recorded_by column

In [52]:
df['recorded_by'].value_counts()

GeoData Consultants Ltd    59400
Name: recorded_by, dtype: int64

The recorded_by column has one value. This will not give any information to our model. So, we dropped it also.

In [53]:
df.drop(columns=['recorded_by'],inplace=True )

### installer column

In [130]:
df['installer'].fillna(value='Unknown',inplace=True)

In [284]:
df['installer'].value_counts().head(100).sum()

49351

In [223]:
df['installer'].replace(to_replace = '0', value ='Unknown' , inplace=True)

In [248]:
df['installer'].replace(to_replace = ('District Water Department', 'District water depar','Distric Water Department'),
                        value ='District water department' , inplace=True)

df['installer'].replace(to_replace = ('FinW','Fini water','FINI WATER'), value ='Fini Water' , inplace=True)
df['installer'].replace(to_replace = 'JAICA', value ='Jaica' , inplace=True)

df['installer'].replace(to_replace = ('COUN', 'District COUNCIL', 'DISTRICT COUNCIL','District Counci', 
                                      'District Council','Council','Counc','District  Council','Distri'),
                                    value ='District council' , inplace=True)

df['installer'].replace(to_replace = ('RC CHURCH', 'RC Churc', 'RC','RC Ch','RC C', 'RC CH','RC church', 
                                      'RC CATHORIC',) , value ='RC Church' , inplace=True)

df['installer'].replace(to_replace = ('Central Government','Tanzania Government',
                                       'central government','Cental Government', 'Cebtral Government', 
                                      'Tanzanian Government','Tanzania government', 'Centra Government' ,
                                      'CENTRAL GOVERNMENT', 'TANZANIAN GOVERNMENT','Central govt', 'Centr', 
                                      'Centra govt') , value ='Central government' , inplace=True)

df['installer'].replace(to_replace = ('World vision', 'World Division','World Vision'),
                                        value ='world vision' , inplace=True)

df['installer'].replace(to_replace = ('Unisef','UNICEF'),value ='Unicef' , inplace=True)
df['installer'].replace(to_replace = 'DANID', value ='DANIDA' , inplace=True)

df['installer'].replace(to_replace = ('villigers', 'villager', 'Villagers', 'Villa', 'Village', 'Villi', 
                                      'Village Council','Village Counil', 'Villages', 'Vill', 'Village community', 
                                      'Villaers', 'Village Community', 'Villag','Villege Council', 'Village council',
                                      'Village  Council','Villagerd', 'Villager', 'Village Technician',
                                      'Village Office','Village community members'),
                                        value ='villagers' , inplace=True)

df['installer'].replace(to_replace =('Commu','Communit','commu','COMMU', 'COMMUNITY') ,
                                        value ='Community' , inplace=True)

df['installer'].replace(to_replace = ('GOVERNMENT', 'GOVER', 'GOVERNME', 'GOVERM','GOVERN','Gover','Gove',
                                      'Governme','Governmen' ) ,value ='Government' , inplace=True)

In [194]:
df['installer'].replace(to_replace = ('Colonial Government') , value ='Colonial government' , inplace=True)
df['installer'].replace(to_replace = ('Government of Misri') , value ='Misri Government' , inplace=True)
df['installer'].replace(to_replace = ('Italy government') , value ='Italian government' , inplace=True)
df['installer'].replace(to_replace = ('British colonial government') , value ='British government' , inplace=True)
df['installer'].replace(to_replace = ('Concern /government') , value ='Concern/Government' , inplace=True)
df['installer'].replace(to_replace = ('Village Government') , value ='Village government' , inplace=True)
df['installer'].replace(to_replace = ('Government and Community') , value ='Government /Community' , inplace=True)
df['installer'].replace(to_replace = ('Cetral government /RC') , value ='RC church/Central Gover' , inplace=True)
df['installer'].replace(to_replace = ('Government /TCRS','Government/TCRS') , value ='TCRS /Government' , inplace=True)
df['installer'].replace(to_replace = ('ADRA /Government') , value ='ADRA/Government' , inplace=True)

In [276]:
(49351*100)/59400 # percentage of seeing value counts of installer

83.08249158249158

We checked the first 100 value counts of installer column. Because we realized some installer names are written with mistake and given as different like 'District Council' and 'District council'. So, we will change the values which we realied. The first 100 value counts contains 49351 values. It means 83% of data is checked.

### funder column

In [285]:
df['funder'].fillna(value='Unknown',inplace=True)
df['funder'].replace(to_replace = '0', value ='Unknown' , inplace=True)

In [286]:
df['funder'].value_counts().head(100)

Government Of Tanzania            9084
Unknown                           4416
Danida                            3114
Hesawa                            2202
Rwssp                             1374
World Bank                        1349
Kkkt                              1287
World Vision                      1246
Unicef                            1057
Tasaf                              877
District Council                   843
Dhv                                829
Private Individual                 826
Dwsp                               811
Norad                              765
Germany Republi                    610
Tcrs                               602
Ministry Of Water                  590
Water                              583
Dwe                                484
Netherlands                        470
Hifab                              450
Adb                                448
Lga                                442
Amref                              425
Fini Water               

### longitude column

In [311]:
df.loc[df['longitude']==0] # to check outside of Tanzania

id             status_group  amount_tsh date_recorded  \
21      6091               functional         0.0    2013-02-10   
53     32376           non functional         0.0    2011-08-01   
168    72678               functional         0.0    2013-01-30   
177    56725           non functional         0.0    2013-01-17   
253    13042  functional needs repair         0.0    2012-10-29   
...      ...                      ...         ...           ...   
59189  62177  functional needs repair         0.0    2011-07-18   
59208   3631               functional         0.0    2013-01-22   
59295  60843  functional needs repair         0.0    2011-07-19   
59324    748               functional         0.0    2013-01-22   
59374  49651               functional         0.0    2012-10-29   

                       funder  gps_height     installer  longitude  \
21                       Dwsp           0           DWE        0.0   
53     Government Of Tanzania           0    Government        0.0   
168                       Wvt           0           WVT        0.0   
177               Netherlands           0           DWE        0.0   
253                    Hesawa           0           DWE        0.0   
...                       ...         ...           ...        ...   
59189                    Dwsp           0           DWE        0.0   
59208                    Dwsp           0           DWE        0.0   
59295                   Rwssp           0           DWE        0.0   
59324            World Vision           0  world vision        0.0   
59374                   Rwssp           0           DWE        0.0   

           latitude                 wpt_name  num_private          basin  \
21    -2.000000e-08                 Muungano            0  Lake Victoria   
53    -2.000000e-08                   Polisi            0  Lake Victoria   
168   -2.000000e-08             Wvt Tanzania            0  Lake Victoria   
177   -2.000000e-08  Kikundi Cha Wakina Mama            0  Lake Victoria   
253   -2.000000e-08                Kwakisusi            0  Lake Victoria   
...             ...                      ...          ...            ...   
59189 -2.000000e-08                   Wazazo            0  Lake Victoria   
59208 -2.000000e-08                  Mtakuja            0  Lake Victoria   
59295 -2.000000e-08                Maendeleo            0  Lake Victoria   
59324 -2.000000e-08                 Mwazwilo            0  Lake Victoria   
59374 -2.000000e-08               Nguvu Kazi            0  Lake Victoria   

       subvillage     region  region_code  district_code      lga  \
21     Ibabachegu  Shinyanga           17              1  Bariadi   
53         Center     Mwanza           19              6    Geita   
168         Ilula  Shinyanga           17              1  Bariadi   
177        Mahaha  Shinyanga           17              1  Bariadi   
253     Nyamatala     Mwanza           19              2     Magu   
...           ...        ...          ...            ...      ...   
59189   Mwamabuli  Shinyanga           17              1  Bariadi   
59208       Mbiti  Shinyanga           17              1  Bariadi   
59295   Mwamalizi  Shinyanga           17              1  Bariadi   
59324       Mbita  Shinyanga           17              1  Bariadi   
59374  Mwamtani A  Shinyanga           17              1  Bariadi   

                    ward  population public_meeting  scheme_name permit  \
21     Ikungulyabashashi           0            NaN          NaN  False   
53           Nyang'hwale           0           True  Nyang'hwale   True   
168            Chinamili           0          False          NaN  False   
177            Bunamhala           0            NaN          NaN  False   
253               Malili           0           True          NaN   True   
...                  ...         ...            ...          ...    ...   
59189             Mhunze           0           True          NaN  False   
59208        Kinang'we

In [312]:
df.loc[df['longitude']!=0].describe()

id     amount_tsh    gps_height     longitude      latitude  \
count  57588.00000   57588.000000  57588.000000  57588.000000  57588.000000   
mean   37106.48807     327.645219    689.325137     35.149669     -5.885572   
std    21454.51421    3043.831403    693.564188      2.607428      2.809876   
min        0.00000       0.000000    -90.000000     29.607122    -11.649440   
25%    18522.75000       0.000000      0.000000     33.285100     -8.643841   
50%    37054.50000       0.000000    426.000000     35.005943     -5.172704   
75%    55667.25000      30.000000   1332.000000     37.233712     -3.372824   
max    74247.00000  350000.000000   2770.000000     40.345193     -0.998464   

        num_private   region_code  district_code    population  \
count  57588.000000  57588.000000   57588.000000  57588.000000   
mean       0.489060     15.217615       5.728311    185.570831   
std       12.426954     17.855254       9.760254    477.744239   
min        0.000000      1.000000       0.000000      0.000000   
25%        0.000000      5.000000       2.000000      0.000000   
50%        0.000000     12.000000       3.000000     35.000000   
75%        0.000000     17.000000       5.000000    230.000000   
max     1776.000000     99.000000      80.000000  30500.000000   

       construction_year  
count       57588.000000  
mean         1997.859919  
std            10.331744  
min          1960.000000  
25%          1995.000000  
50%          2000.000000  
75%          2004.000000  
max          2013.000000

In [314]:
df['longitude'].replace(to_replace = 0 , value =35.15, inplace=True)

In [327]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 27 columns):
status_group             59400 non-null object
amount_tsh               59400 non-null float64
date_recorded            59400 non-null object
funder                   59400 non-null object
gps_height               59400 non-null int64
installer                59400 non-null object
longitude                59400 non-null float64
latitude                 59400 non-null float64
num_private              59400 non-null int64
basin                    59400 non-null object
subvillage               59029 non-null object
region                   59400 non-null object
region_code              59400 non-null int64
district_code            59400 non-null int64
lga                      59400 non-null object
ward                     59400 non-null object
population               59400 non-null int64
public_meeting           56066 non-null object
permit                   56344 non-null object


### wpt_name & scheme_name columns

In [316]:
df['wpt_name'].value_counts()

none                   3563
Shuleni                1748
Zahanati                830
Msikitini               535
Kanisani                323
                       ... 
Msese                     1
Kwa Kichumvi              1
Yope                      1
Kwa Mzee Nzengahata       1
Kasimbagu                 1
Name: wpt_name, Length: 37400, dtype: int64

In [319]:
df['scheme_name'].value_counts()

K                           682
None                        644
Borehole                    546
Chalinze wate               405
M                           400
                           ... 
Rain Water Harvesting         1
ADP Simbu                     1
Mradi wa maji wa Rusimbi      1
Bulenya water supply          1
KAIBANJA PUMPING SCHEME       1
Name: scheme_name, Length: 2696, dtype: int64

In [317]:
df.drop(columns=['wpt_name','scheme_name','id','region_code'],inplace=True )

In [325]:
df.duplicated(subset='id', keep='first').sum()

0

In [329]:
df['region'].value_counts()

Iringa           5294
Shinyanga        4982
Mbeya            4639
Kilimanjaro      4379
Morogoro         4006
Arusha           3350
Kagera           3316
Mwanza           3102
Kigoma           2816
Ruvuma           2640
Pwani            2635
Tanga            2547
Dodoma           2201
Singida          2093
Mara             1969
Tabora           1959
Rukwa            1808
Mtwara           1730
Manyara          1583
Lindi            1546
Dar es Salaam     805
Name: region, dtype: int64

### amount_tsh column

In [60]:
df['amount_tsh'].value_counts()

0.0         41639
500.0        3102
50.0         2472
1000.0       1488
20.0         1463
            ...  
8500.0          1
6300.0          1
220.0           1
138000.0        1
12.0            1
Name: amount_tsh, Length: 98, dtype: int64

In [62]:
df.loc[df['amount_tsh']==0].groupby('status_group').count()

id  amount_tsh  date_recorded  funder  gps_height  \
status_group                                                                    
functional               19706       19706          19706   18026       19706   
functional needs repair   3048        3048           3048    2635        3048   
non functional           18885       18885          18885   17697       18885   

                         installer  longitude  latitude  wpt_name  \
status_group                                                        
functional                   18012      19706     19706     19706   
functional needs repair       2633       3048      3048      3048   
non functional               17698      18885     18885     18885   

                         num_private  basin  subvillage  region  region_code  \
status_group                                                                   
functional                     19706  19706       19501   19706        19706   
functional needs repair         3048   3048        3047    3048         3048   
non functional                 18885  18885       18720   18885        18885   

                         district_code    lga   ward  population  \
status_group                                                       
functional                       19706  19706  19706       19706   
functional needs repair           3048   3048   3048        3048   
non functional                   18885  18885  18885       18885   

                         public_meeting  scheme_name  permit  \
status_group                                                   
functional                        18444         8759   18374   
functional needs repair            2906         1286    2775   
non functional                    17475         8619   17862   

                         construction_year  extraction_type_group  management  \
status_group                                                                    
functional                           19706                  19706       19706   
functional needs repair               3048                   3048        3048   
non functional                       18885                  18885       18885   

                         payment  water_quality  quantity  source  \
status_group                                                        
functional                 19706          19706     19706   19706   
functional needs repair     3048           3048      3048    3048   
non functional             18885          18885     18885   18885   

                         waterpoint_type  
status_group                              
functional                         19706  
functional needs repair             3048  
non functional                     18885

In [63]:
df['amount_tsh'].value_counts()/df['amount_tsh'].count()

0.0         0.700993
500.0       0.052222
50.0        0.041616
1000.0      0.025051
20.0        0.024630
              ...   
8500.0      0.000017
6300.0      0.000017
220.0       0.000017
138000.0    0.000017
12.0        0.000017
Name: amount_tsh, Length: 98, dtype: float64

# We will decided this zero values later?????

### gps_height column

In [65]:
df['gps_height'].value_counts()

 0       20438
-15         60
-16         55
-13         55
-20         52
         ...  
 2285        1
 2424        1
 2552        1
 2413        1
 2385        1
Name: gps_height, Length: 2428, dtype: int64

In [71]:
df['gps_height'].value_counts()/df['gps_height'].count()

 0       0.344074
-15      0.001010
-16      0.000926
-13      0.000926
-20      0.000875
           ...   
 2285    0.000017
 2424    0.000017
 2552    0.000017
 2413    0.000017
 2385    0.000017
Name: gps_height, Length: 2428, dtype: float64

### population column

In [73]:
df['population'].value_counts()

0       21381
1        7025
200      1940
150      1892
250      1681
        ...  
3241        1
1960        1
1685        1
2248        1
1439        1
Name: population, Length: 1049, dtype: int64

In [74]:
df.loc[df['population']==0].groupby('status_group').count()

id  amount_tsh  date_recorded  funder  gps_height  \
status_group                                                                    
functional               11274       11274          11274   10048       11274   
functional needs repair   1775        1775           1775    1464        1775   
non functional            8332        8332           8332    7753        8332   

                         installer  longitude  latitude  wpt_name  \
status_group                                                        
functional                   10033      11274     11274     11274   
functional needs repair       1462       1775      1775      1775   
non functional                7755       8332      8332      8332   

                         num_private  basin  subvillage  region  region_code  \
status_group                                                                   
functional                     11274  11274       11071   11274        11274   
functional needs repair         1775   1775        1775    1775         1775   
non functional                  8332   8332        8174    8332         8332   

                         district_code    lga   ward  population  \
status_group                                                       
functional                       11274  11274  11274       11274   
functional needs repair           1775   1775   1775        1775   
non functional                    8332   8332   8332        8332   

                         public_meeting  scheme_name  permit  \
status_group                                                   
functional                        10700         4303   10596   
functional needs repair            1732          573    1613   
non functional                     7967         3479    8063   

                         construction_year  extraction_type_group  management  \
status_group                                                                    
functional                           11274                  11274       11274   
functional needs repair               1775                   1775        1775   
non functional                        8332                   8332        8332   

                         payment  water_quality  quantity  source  \
status_group                                                        
functional                 11274          11274     11274   11274   
functional needs repair     1775           1775      1775    1775   
non functional              8332           8332      8332    8332   

                         waterpoint_type  
status_group                              
functional                         11274  
functional needs repair             1775  
non functional                      8332